votingsystem/src/voting.rs

```diff
@@ -0,0 +1,419 @@
+                let corresponding_paper_id = vote_paper.vote_paper_id;
+                let vote = Vote::new(vote_paper_id, choice, voter_signature);
+                let vote_id = vote.clone().vote_id;
+                vote_paper.set_used_in(vote_id.clone());
+                let meeting_id = vote_paper.get_general_meeting_id();
+                self.storage_mut().set(vote_id.as_ref(), serde_cbor::to_vec(&vote).unwrap());
+                let box_key = crate::generate_key(&meeting_id);
+                let mut vote_box = self.meeting_mut().get_box(box_key.as_slice()).unwrap();
+                vote_box.drop_in_box(vote.clone().vote_id);
+                Ok(vote_id.id)
+            }
+        }
+    }
+}
+
+impl Stateful for Context {
+    fn set_storage(&mut self, storage: ServiceRef<dyn SubStorageAccess>) {
+        self.storage.replace(storage.unwrap_import().into_remote());
+    }
+}
+
+impl TxOwner for Context {
+    fn block_opened(&mut self, _: &Header) -> Result<(), HeaderError> {
+        Ok(())
+    }
+
+    fn execute_transaction(&mut self, transaction: &Transaction) -> Result<TransactionOutcome, ()> {
+        if let Err(error) = self.excute_tx(transaction) {
+            match error {
+                ExecuteError::InvalidMetadata => Err(()),
+                ExecuteError::InvalidSign => Err(()),
+                ExecuteError::InvalidFormat => Err(()),
+                ExecuteError::InvalidAdmin => Err(()),
+                ExecuteError::UsedVotePaper => Err(()),
+                ExecuteError::InvalidNumberofAgendas => Err(()),
+                ExecuteError::InvalidMeeting => Err(()),
+                ExecuteError::InvalidMeetingTime => Err(()),
+                ExecuteError::InvalidVotingTime => Err(()),
+                ExecuteError::InvalidChoice => Err(()),
+            }
+        } else {
+            Ok(Default::default())
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "Vote");
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+        tx.verify().map_err(|_| todo_fixthis)?;
+
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                voter_name,
+                number_of_shares,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting =
+                    serde_cbor::from_slice(&self.storage().get(general_meeting_id.as_ref()).unwrap()).unwrap();
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting).map_err(|_| todo_fixthis)
+                }
+                if meeting.number_of_agendas < agenda_number {
```

Comment:
Does the agenda number start from 0? or 1?

If it starts from 0, this condition is wrong. The condition should be `meeting.number_of_agendas <= agenda_number`.
If it starts from 1, this condtion is wrong. The condition should be `0 < agenda_number && agenda_number <= meeting.number_of_agendas`.